libportability-gfx/src/impls.rs

```diff
@@ -280,31 +272,118 @@ pub extern "C" fn gfxGetPhysicalDeviceFormatProperties(
+    }
+}
+#[inline]
+pub extern "C" fn gfxGetPhysicalDeviceImageFormatProperties2KHR(
+    adapter: VkPhysicalDevice,
+    pImageFormatInfo: *const VkPhysicalDeviceImageFormatInfo2KHR,
+    pImageFormatProperties: *mut VkImageFormatProperties2KHR,
+) -> VkResult {
+    let mut properties = None;
+
+    let mut ptr = pImageFormatInfo as *const VkStructureType;
+    while !ptr.is_null() {
+        match unsafe { *ptr } {
+            VkStructureType::VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_IMAGE_FORMAT_INFO_2_KHR => {
+                let data = unsafe {
+                    (ptr as *const VkPhysicalDeviceImageFormatInfo2KHR).as_ref().unwrap()
+                };
+                properties = get_physical_device_image_format_properties(adapter, data);
+            }
+            VkStructureType::VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_IMAGE_VIEW_SUPPORT_EXTX => {
+                let data = unsafe {
+                    (ptr as *const VkPhysicalDeviceImageViewSupportEXTX).as_ref().unwrap()
+                };
+                //TODO: provide the data from gfx-rs itself
+                // copied from `map_format_with_swizzle`
+                let identity = VkComponentMapping {
+                    r: VkComponentSwizzle::VK_COMPONENT_SWIZZLE_R,
+                    g: VkComponentSwizzle::VK_COMPONENT_SWIZZLE_G,
+                    b: VkComponentSwizzle::VK_COMPONENT_SWIZZLE_B,
+                    a: VkComponentSwizzle::VK_COMPONENT_SWIZZLE_A,
+                };
+                let bgra = VkComponentMapping {
+                    r: identity.b,
+                    b: identity.r,
+                    .. identity
+                };
+                if data.components != identity && cfg!(feature = "gfx-backend-metal") {
+                    let supported = match data.format {
+                        VkFormat::VK_FORMAT_R8_UNORM => data.components == VkComponentMapping {
+                            r: VkComponentSwizzle::VK_COMPONENT_SWIZZLE_ZERO,
+                            g: VkComponentSwizzle::VK_COMPONENT_SWIZZLE_ZERO,
+                            b: VkComponentSwizzle::VK_COMPONENT_SWIZZLE_ZERO,
+                            a: VkComponentSwizzle::VK_COMPONENT_SWIZZLE_R,
+                        },
+                        VkFormat::VK_FORMAT_R8G8B8A8_UNORM => data.components == bgra,
+                        VkFormat::VK_FORMAT_B8G8R8A8_UNORM => data.components == bgra,
+                        VkFormat::VK_FORMAT_B8G8R8A8_SRGB => data.components == bgra,
+                        VkFormat::VK_FORMAT_B5G6R5_UNORM_PACK16 => data.components == bgra,
+                        _ => false,
+                    };
+                    if !supported {
+                        return VkResult::VK_ERROR_FORMAT_NOT_SUPPORTED;
+                    }
+                }
+            }
+            other => {
+                warn!("Unrecognized {:?}, skipping", other);
+            }
+        };
+        ptr = unsafe {
```

Comment:
In the future we should probably consider move some of the `pNext` chaining logic into a more explicit function somehow, because we'll have to use this in a few places anyway

Reply:
yeah, agreed. Need some sort of a safe-ish iterator wrapper for those